consensus/istanbul/backend/engine.go

```diff
@@ -175,20 +177,112 @@ func (sb *Backend) verifyCascadingFields(chain consensus.ChainReader, header *ty
 func (sb *Backend) VerifyHeaders(chain consensus.ChainReader, headers []*types.Header, seals []bool) (chan<- struct{}, <-chan error) {
 	abort := make(chan struct{})
 	results := make(chan error, len(headers))
-	go func() {
-		for i, header := range headers {
-			err := sb.verifyHeader(chain, header, headers[:i])
 
-			select {
-			case <-abort:
-				return
-			case results <- err:
+	// If we are in "lightest" mode, we do batch verification of all BLS signatures
+	// in the headers, so there is no need to iterate over the blocks
+	if chain.Config().FullHeaderChainAvailable {
+		go func() {
+			for i, header := range headers {
+				err := sb.verifyHeader(chain, header, headers[:i])
+
+				select {
+				case <-abort:
+					return
+				case results <- err:
+				}
 			}
-		}
-	}()
+		}()
+	} else {
+		go func() {
+			err := sb.batchVerifyHeaders(chain, headers)
+			// even though we get 1 return value from the batched verification
+			// method, we need to fill the whole channel of results with return values
+			for i := 0; i < len(headers); i++ {
+				select {
+				case <-abort:
+					return
+				case results <- err:
+				}
+			}
+		}()
+	}
 	return abort, results
 }
 
+// batchVerifyHeaders is called in ultralight mode and performs batch BLS signature
+// verification over all provided headers and their corresponding validator sets
+func (sb *Backend) batchVerifyHeaders(chain consensus.ChainReader, headers []*types.Header) error {
```

Comment:
This seems to have some duplicated code from `verifyHeader`, see my suggestion above for how this might be shared